src/test/java/com/microsoft/sqlserver/jdbc/connection/TimeoutTest.java

```diff
@@ -29,66 +30,124 @@
         long timerEnd = 0;
 
-        timerStart = System.currentTimeMillis();
+        long timerStart = System.currentTimeMillis();
         // Try a non existing server and see if the default timeout is 15 seconds
         try (Connection con = PrepUtil.getConnection("jdbc:sqlserver://" + randomServer + ";user=sa;password=pwd;")) {
+            fail(TestResource.getResource("R_shouldNotConnect"));
+        } catch (Exception e) {
+            assertTrue(e.getMessage().contains(TestResource.getResource("R_tcpipConnectionToHost")));
+            timerEnd = System.currentTimeMillis();
+        }
+
+        verifyTimeout(timerEnd - timerStart, defaultTimeout);
+    }
 
+    @Test
+    public void testURLLoginTimeout() {
+        long timerEnd = 0;
+        int timeout = 10;
+
+        long timerStart = System.currentTimeMillis();
+
+        try (Connection con = PrepUtil
+                .getConnection("jdbc:sqlserver://" + randomServer + ";user=sa;password=pwd;logintimeout=" + timeout)) {
+            fail(TestResource.getResource("R_shouldNotConnect"));
         } catch (Exception e) {
             assertTrue(e.getMessage().contains(TestResource.getResource("R_tcpipConnectionToHost")));
             timerEnd = System.currentTimeMillis();
         }
 
-        assertTrue(0 != timerEnd, TestResource.getResource("R_shouldNotConnect"));
+        verifyTimeout(timerEnd - timerStart, timeout);
+    }
+
+    @Test
+    public void testDMLoginTimeoutApplied() {
+        long timerEnd = 0;
+        int timeout = 10;
+
+        DriverManager.setLoginTimeout(timeout);
+        long timerStart = System.currentTimeMillis();
+
+        try (Connection con = PrepUtil.getConnection("jdbc:sqlserver://" + randomServer + ";user=sa;password=pwd")) {
+            fail(TestResource.getResource("R_shouldNotConnect"));
+        } catch (Exception e) {
+            assertTrue(e.getMessage().contains(TestResource.getResource("R_tcpipConnectionToHost")));
+            timerEnd = System.currentTimeMillis();
+        }
 
-        long timeDiff = timerEnd - timerStart;
-        assertTrue(timeDiff > 14000);
+        verifyTimeout(timerEnd - timerStart, timeout);
+    }
+
+    @Test
+    public void testDMLoginTimeoutNotApplied() {
+        long timerEnd = 0;
+        int timeout = 10;
+
+        DriverManager.setLoginTimeout(timeout * 3); // 30 seconds
+        long timerStart = System.currentTimeMillis();
+
+        try (Connection con = PrepUtil
+                .getConnection("jdbc:sqlserver://" + randomServer + ";user=sa;password=pwd;loginTimeout=" + timeout)) {
+            fail(TestResource.getResource("R_shouldNotConnect"));
+        } catch (Exception e) {
+            assertTrue(e.getMessage().contains(TestResource.getResource("R_tcpipConnectionToHost")));
+            timerEnd = System.currentTimeMillis();
+        }
+
+        verifyTimeout(timerEnd - timerStart, timeout);
+
+        DriverManager.setLoginTimeout(0); // Default to 0 again
```

Comment:
Please add these 2 lines to a `finally` block to make sure DM timeout is set back to its default value even if the test fails.